### PR TITLE
Adding TGB negative sampling hook 

### PR DESCRIPTION
**Pull request by shenyangHuang**
- [ ] unified all split-mode for validation set to be called `val` instead of `valid` for unification
- [ ] added `TGBNegativeEdgeSamplerHook` in `hooks.py`
- [ ] added TGB example for edgebank, getting the same performance, dataloader and negative sampling hook works
- [ ] added test for `TGBNegativeEdgeSamplerHook`

To do 
- [ ] add TGAT and TGN example with TGB hook, maybe that goes in a different PR

### CODE REVIEW

File: examples/linkproppred/TGB/edgebank.py
Comment (Jacob-Chmura) [inline examples/linkproppred/TGB/edgebank.py:1]: I'm wondering whether we should have seperate examples for tgb/non-tgb. Or just have everything as tgb (if possible). What are your thoughts? My view is that almost everyone would be doing tgb eval anyway, and it's a bit simpler to have a single example for each method.

File: examples/linkproppred/TGB/edgebank.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+import argparse
+
+import numpy as np
+import torch
+from tgb.linkproppred.dataset_pyg import PyGLinkPropPredDataset
```
Comment (Jacob-Chmura) [inline examples/linkproppred/TGB/edgebank.py:5]: TGB is now a dependency for the examples, so it should be added to the examples group: `uv add --group examples tgb`

File: examples/linkproppred/TGB/edgebank.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            query_dst = torch.cat([torch.tensor([batch.dst[idx]]), neg_batch])
+
+            y_pred = model(query_src, query_dst)
+            # compute MRR
+            input_dict = {
+                'y_pred_pos': np.array([y_pred[0]]),
+                'y_pred_neg': np.array(y_pred[1:]),
+                'eval_metric': [eval_metric],
+            }
+            perf_list.append(evaluator.eval(input_dict)[eval_metric])
```
Comment (Jacob-Chmura) [inline examples/linkproppred/TGB/edgebank.py:56]: Is this worth vectorizing?

File: examples/linkproppred/TGB/edgebank.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        model.update(batch.src, batch.dst, batch.time)
+    metric_dict = {}
+    metric_dict[eval_metric] = float(np.mean(perf_list))
+    return metric_dict
+
+
+args = parser.parse_args()
+seed_everything(args.seed)
+
+# * setting up tgb neg sampler
```
Comment (Jacob-Chmura) [inline examples/linkproppred/TGB/edgebank.py]:
```suggestion
```

File: examples/linkproppred/TGB/edgebank.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+dataset = PyGLinkPropPredDataset(name=args.dataset, root='datasets')
+eval_metric = dataset.eval_metric
+neg_sampler = dataset.negative_sampler
+evaluator = Evaluator(name=args.dataset)
+dataset.load_val_ns()
+dataset.load_test_ns()
```
Comment (Jacob-Chmura) [inline examples/linkproppred/TGB/edgebank.py:71]: This is not a huge deal, but it seems not ideal that we (i.e. the user) needs to load in a completely separate copy of the data to initialize the negative sampler. Is there another API in tgb to construct the negative sampler directly?

If not, what do you think about hiding this behind `TBNegativeSamplerHook`: specifically, just let the negative sampler hook take a dataset name (`args.dataset`)?

Comment (Jacob-Chmura) [reply examples/linkproppred/TGB/edgebank.py:71]: It's nice that the evaluator just takes the dataset name. Presumably `dataset.load_ns()` must be called by the user directly? I'm asking because we don't actually use the dataset explicitly. The only thing that actually keeps reference to it is the negative sampler, but we do `load_ns` _after_ that.

Basically, do we need the `load_ns` and if so, how does it interact with the `tgm` notion of time granularity. Can it be unified?

File: tgm/hooks.py [hunk 5]
```diff
@@ -123,6 +124,48 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
         return batch
 
 
+class TGBNegativeEdgeSamplerHook:
+    requires: Set[str] = set()
+    produces = {'neg', 'neg_batch_list'}
+    r"""Load data from DGraph using pre-generated TGB negative samples.
+    Make sure to perform `dataset.load_val_ns()` or `dataset.load_test_ns()` before using this hook.
```
Comment (Jacob-Chmura) [inline tgm/hooks.py:131]: Thoughts on taking dataset name as string argument, and managing the ns load ourselves without the user having to remember to do it?

File: tgm/hooks.py [hunk 6]
```diff
@@ -123,6 +124,48 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
...
+    Args:
```
Comment (Jacob-Chmura) [inline tgm/hooks.py:133]:
```suggestion
    
    Args:
```

File: tgm/hooks.py [hunk 7]
```diff
@@ -123,6 +124,48 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
...
+        if neg_sampler.eval_set[split_mode] is None:  # type: ignore
+            raise ValueError(
+                f'please run load_{split_mode}_ns() before using this hook'
+            )
+
+        self.neg_sampler = neg_sampler
+        self.split_mode = split_mode
+
+    def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
+        # this might complain if the edge is not found in the negative sampler, which could happen if the user is not using the correct version of dataset
```
Comment (Jacob-Chmura) [inline tgm/hooks.py:155]: By complain, you mean raise an error? Or warning? Can we recover from the exception if it occurs?

File: tgm/hooks.py [hunk 8]
```diff
@@ -123,6 +124,48 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
...
+        neg_batch_list = self.neg_sampler.query_batch(  # type: ignore
+            batch.src, batch.dst, batch.time, split_mode=self.split_mode
+        )
+        queries = []
+        tensor_batch_list = []
+        for _, neg_batch in enumerate(neg_batch_list):
```
Comment (Jacob-Chmura) [inline tgm/hooks.py]:
```suggestion
        for neg_batch in neg_batch_list:
```

File: tgm/hooks.py [hunk 9]
```diff
@@ -123,6 +124,48 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
...
+            queries.append(neg_batch)
+            tensor_batch_list.append(torch.tensor(neg_batch, dtype=torch.long))
+        unique_neg = np.unique(np.concatenate(queries))
+        batch.neg = torch.tensor(unique_neg, dtype=torch.long)  # type: ignore
```
Comment (Jacob-Chmura) [inline tgm/hooks.py:165]: This isn't used in the current example. Do we need this here?

File: test/test_hooks/test_tgb_negative_sampling_hook.py
Comment (Jacob-Chmura) [inline test/test_hooks/test_tgb_negative_sampling_hook.py:1]: Nice

File: examples/linkproppred/TGB/edgebank.py
Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:1]: Thanks Jacob, I think going forward, we can prioritize having TGB evaluation. However, it is also possible that people might want to do their own evaluation and the basic example work like that. As this is a software library, it is always good to have more examples. I think we restructure the example (also easily) when we go with alpha release and when integrating tutorials, not an issue for now I think. 

TLDR: currently and for clarity, let's keep the TGB folder (also it requires TGB dependency). We can restructure for the alpha release

File: examples/linkproppred/TGB/edgebank.py (hunk 1 above)
Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:5]: `uv add --group examples py-tgb`

Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:5]: changed in `uv.lock` to 

```
examples = [
    { name = "torchmetrics" },
    { name = "tqdm" },
    { name = "py-tgb" },
]
```

Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:5]: ran the above and added to `uv.lock`

Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:5]: also added `uv sync --group examples` to README

File: examples/linkproppred/TGB/edgebank.py (hunk 2 above)
Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:56]: what you mean? you mean `np.array`, you have to do that for standardized input to TGB evaluator. Note that the `evaluator` now is from TGB so it handles all the evaluation logic. This is required

File: examples/linkproppred/TGB/edgebank.py (hunk 4 above)
Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:71]: yes and no, by default the dataset in TGB doesn't `load_ns` because some of the larger dataset has very large `ns` size (ten times larger than dataset) so if the user don't do link prediction, they don't need ns.

Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:71]: Here is the TLDR trade-off:
- solution one: each TGB hook loads its own dataset, calls load_ns and get its own neg sampler
problem: actually we just need one neg_sampler, this approach will create two copies of dataset, each for val and test, benefit: more straightforward, more abstracted away

- solution two: current approach, each TGB hook gets an input neg_sampler that allows it to get the test and val samples. one copy of dataset in example

Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:71]: I think solution one is nicer and more abstracted away and the user think less of what it is doing under the hood. However, I implemented solution 2 because:
- it avoid loading one copy of dataset for each TGB hook
- it makes the negative sampler hook very general actually, as long as you are passing a neg_sampler (similarly setup as TGB one), you can reuse this in your own custom dataset. You would still call `TGBNegativeEdgeSamplerHook` but just make your own neg_sampler, this is quite helpful especially when people want to try their custom negative sample etc. 
- in both cases, the evaluation loop is the same, so I don't think it makes a fundamental difference, user will just copy what we did in TGB example for the most part


TLDR: do we want `TGBNegativeEdgeSamplerHook` to be able to take in custom negative sampler? or do we want to fix it only for TGB? My personal take is that TGB evaluation is simply a novel / more robust way of evaluation and the user can also use their own negative sampler on novel datasets

Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:71]: I suggest to keep it as is but happy to discuss more.

File: tgm/hooks.py (hunk 7 above)
Comment (shenyangHuang) [reply tgm/hooks.py:155]: it should throw an error, if you can't find the negative sample of an edge, it could be:
- you are running the wrong version of the dataset
- you didn't load your test or val ns (and TGB will complain correspondingly)

both errors will be throw by the neg_sampler from TGB

File: tgm/hooks.py (hunk 9 above)
Comment (shenyangHuang) [reply tgm/hooks.py:165]: yes we need this for neighbor sampler :)

File: README.md [hunk 10]
```diff
@@ -49,6 +49,12 @@ uv pip install git+https://github.com/tgm-team/tgm.git
 python -c 'import tgm; print(tgm.__version__)'
 ```
 
+running examples
```
Comment (Jacob-Chmura) [inline README.md:52]: Sure, we can just keep this here for temporary purposes, but this will be refined prior to release.

File: examples/linkproppred/TGB/edgebank.py
Comment (Jacob-Chmura) [reply examples/linkproppred/TGB/edgebank.py:1]: Sounds good.

File: tgm/hooks.py (hunk 7 above)
Comment (Jacob-Chmura) [reply tgm/hooks.py:155]: Okay, so safe for use to ignore I suppose (unless we handle the `load_ns` ourselves)

File: examples/linkproppred/TGB/edgebank.py (hunk 2 above)
Comment (Jacob-Chmura) [reply examples/linkproppred/TGB/edgebank.py:56]: What I mean is, do a single batched inference for the entire batch as opposed to looping through the neg_batch_list one at a time, doing all the inferences, and calling eval separately on each negative sub batch. 

I understand that TGB evaluator expects ndarray, I'm wondering if we can just do a single inference, single `to_numpy()` , then single eval on the batched np array?

File: examples/linkproppred/TGB/edgebank.py (hunk 4 above)
Comment (Jacob-Chmura) [reply examples/linkproppred/TGB/edgebank.py:71]: Okay. Generally I think the `TGBHook` should be specialized for TGB. And users can create there own hook that's more generic. If I were to use the TGB hook using my own negative sampler, how do I know what API this hook expects? How do I know the exceptions it could raise etc? 

The avoiding of loading multiple copies across test/val/train hooks is valid. Overall, I'm happy to leave this as is, and we can revisit it later. I antipicate that at some point we will make it such that the hook manager can avoid these duplicate loads. 

I also think it's a design from TGB side that I need to understand more.

**Review by Jacob-Chmura — APPROVED** [review]

File: examples/linkproppred/TGB/edgebank.py (hunk 2 above)
Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:56]: Hi Jacob, I see what you mean. Looping through each positive edge and its negative edges are how it was setup in TGB and OGB, conceptually simple. Yes, you can just get a stack of tensors for evaluation as well, requires changing the TGB code to work with batch. Definitely might have improvements (at least eliminating some of the loops) but out of scope for this PR. Also if there are many edges in a batch, stacking them all into a tensor might cause OOM (say batch size = 1,000,000), though it is unrealistic in the current evaluation

Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:56]: good point to maybe leave it as optimization in the future. Requires TGB code change

Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:56]: https://github.com/shenyangHuang/TGB/blob/main/tgb/linkproppred/evaluate.py Here is pointer to TGB evaluator

File: examples/linkproppred/TGB/edgebank.py (hunk 4 above)
Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:71]: Yeah I think avoiding duplicate loads in 
- hook manager 
- dataloader 
is an issue we should address

File: examples/linkproppred/TGB/edgebank.py (hunk 2 above)
Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:56]: See issue #31

File: examples/linkproppred/TGB/edgebank.py (hunk 4 above)
Comment (shenyangHuang) [reply examples/linkproppred/TGB/edgebank.py:71]: I created an issue to keep track, see #30

File: README.md (hunk 10 above)
Comment (shenyangHuang) [reply README.md:52]: yeah, we can keep it for development build and most people will want to run the examples anyways
